omnigibson/transition_rules.py

```diff
@@ -296,358 +710,453 @@ def transition(self, individual_objects, group_objects):
-        for obj_category, objs in group_objects.items():
-            inside_objs = []
-            for obj in objs:
-                if obj.states[Inside].get_value(blender):
-                    inside_objs.append(obj)
-            # Make sure the number of objects inside meets the required threshold, else we trigger a failure
-            if len(inside_objs) < self.obj_requirements[obj_category]:
+            recipe (dict): Recipe whose systems should be checked
+            container (BaseObject): Container object that should contain all of @recipe's input systems
+
+        Returns:
+            bool: True if all the input systems are contained
+        """
+        for system_name in recipe["input_systems"]:
+            if not container.states[Contains].get_value(system=get_system(system_name=system_name)):
                 return False
-            # We mutate the group_objects in place so that we only keep the ones inside the blender
-            group_objects[obj_category] = inside_objs
+        return True
 
-        # Check whether we have sufficient physical particles as well
-        for system_name, n_min_particles in self.particle_requirements.items():
-            if not is_system_active(system_name):
+    @classmethod
+    def _validate_nonrecipe_systems_not_contained(cls, recipe, container):
+        """
+        Validates whether all systems not relevant to @recipe are not contained in @container
+
+        Args:
+            recipe (dict): Recipe whose systems should be checked
+            container (BaseObject): Container object that should contain all of @recipe's input systems
+
+        Returns:
+            bool: True if none of the non-relevant systems are contained
+        """
+        relevant_systems = set(recipe["input_systems"])
+        for system in PhysicalParticleSystem.get_active_systems():
+            if system.name not in relevant_systems and container.states[Contains].get_value(system=system):
                 return False
+        return True
 
-            system = get_system(system_name)
-            if system.n_particles == 0:
+    @classmethod
+    def _validate_recipe_objects_are_contained(cls, recipe, in_volume):
+        """
+        Validates whether @recipe's input_objects are all contained in the container represented by @in_volume
+
+        Args:
+            recipe (dict): Recipe whose objects should be checked
+            in_volume (n-array): (N,) flat boolean array corresponding to whether every object from
+                cls._OBJECTS is inside the corresponding container
+
+        Returns:
+            bool: True if all the input object quantities are contained
+        """
+        for obj_category, obj_quantity in recipe["input_objects"].items():
+            if np.sum(in_volume[cls._CATEGORY_IDXS[obj_category]]) < obj_quantity:
                 return False
+        return True
+
+    @classmethod
+    def _validate_nonrecipe_objects_not_contained(cls, recipe, in_volume):
```

Comment:
Based on our yesterday's discussion, maybe we can relax this constraint for `MixingWandRule`?

Reply:
- assert no input objects in mixing wand rule
- allow non-relevant objects to be in the container for mixing wand rule
    - do this by overlap_sphere in generate_particles_from_link call, and make the filtering a utility function
- handle visual particle systems as inputs as well, check for "CoveredAny"